src/QMCWaveFunctions/Fermion/SlaterDetBuilder.cpp

```diff
@@ -281,65 +268,36 @@ std::unique_ptr<WaveFunctionComponent> SlaterDetBuilder::buildComponent(xmlNodeP
             std::make_unique<SPOSetProxyForMSD>(std::move(spo_clones[0]), targetPtcl.first(0), targetPtcl.last(0));
         auto spo_dn =
             std::make_unique<SPOSetProxyForMSD>(std::move(spo_clones[1]), targetPtcl.first(1), targetPtcl.last(1));
-        if (UseBackflow)
+        if (BFTrans)
         {
           app_summary() << "    Using backflow transformation." << std::endl;
-          multislaterdet_0 = std::make_unique<MultiSlaterDeterminantWithBackflow>(targetPtcl, std::move(spo_up),
-                                                                                  std::move(spo_dn), BFTrans);
-          createMSD(*multislaterdet_0, cur);
+          auto msd_all_dets = std::make_unique<MultiSlaterDeterminantWithBackflow>(targetPtcl, std::move(spo_up),
+                                                                                   std::move(spo_dn), BFTrans);
+          createMSD(*msd_all_dets, cur, BFTrans);
+          msd_all_dets->setBF(BFTrans);
+          built_singledet_or_multidets = std::move(msd_all_dets);
         }
         else
         {
-          multislaterdet_0 = std::make_unique<MultiSlaterDeterminant>(targetPtcl, std::move(spo_up), std::move(spo_dn));
-          createMSD(*multislaterdet_0, cur);
+          auto msd_all_dets =
+              std::make_unique<MultiSlaterDeterminant>(targetPtcl, std::move(spo_up), std::move(spo_dn));
+          createMSD(*msd_all_dets, cur, BFTrans);
+          built_singledet_or_multidets = std::move(msd_all_dets);
         }
       }
     }
     cur = cur->next;
   }
 
 
-  if (!multiDet && !slaterdet_0)
-  {
-    //fatal
-    PRE.error("Failed to create a SlaterDeterminant.", true);
-    return nullptr;
-  }
-  if (multiDet && (!multislaterdet_0 && !multislaterdetfast_0))
+  if (built_singledet_or_multidets)
+    return built_singledet_or_multidets;
+  else
   {
     //fatal
-    PRE.error("Failed to create a MultiSlaterDeterminant.", true);
+    PRE.error("Failed to create a Single Slater determinant or Multi Slater determinants object.", true);
```

Comment:
Then there is this alternative PRE.error mechanism for fatal errors which then returns a nullptr.  As far as I can tell that will eventually cause an exception to be thrown in TrialWaveFunction::addComponent when a operator-> on the unique pointer returned results in rubbish.  The PRE.error message is lost. It would be much better to have an exception that could be trapped in the debugger at the site things went wrong. This would require more time to figure out.  For a fatal error just throw an appropriate exception where it occurs. If its a problem all ranks will encounter throw a UniformCommunicateError.

Reply:
ProgressReportEngine should go away as #3457